alg/CATENets/catenets/models/base.py

```diff
@@ -166,75 +223,107 @@ def fit_and_select_params(self, X, y, w, p=None, param_grid: dict = None):
-                          penalty_l2: float = DEFAULT_PENALTY_L2,
-                          step_size: float = DEFAULT_STEP_SIZE,
-                          n_iter: int = DEFAULT_N_ITER, batch_size: int = DEFAULT_BATCH_SIZE,
-                          val_split_prop: float = DEFAULT_VAL_SPLIT, early_stopping: bool = True,
-                          patience: int = DEFAULT_PATIENCE, n_iter_min: int = DEFAULT_N_ITER_MIN,
-                          verbose: int = 1, n_iter_print: int = DEFAULT_N_ITER_PRINT,
-                          seed: int = DEFAULT_SEED, return_val_loss: bool = False,
-                          nonlin: str = DEFAULT_NONLIN, avg_objective: bool = False):
+        return self
+
+
+def train_output_net_only(
+    X: jnp.ndarray,
+    y: jnp.ndarray,
+    binary_y: bool = False,
+    n_layers_out: int = DEFAULT_LAYERS_OUT,
+    n_units_out: int = DEFAULT_UNITS_OUT,
+    n_layers_r: int = 0,
+    n_units_r: int = DEFAULT_UNITS_R,
+    penalty_l2: float = DEFAULT_PENALTY_L2,
+    step_size: float = DEFAULT_STEP_SIZE,
+    n_iter: int = DEFAULT_N_ITER,
+    batch_size: int = DEFAULT_BATCH_SIZE,
+    val_split_prop: float = DEFAULT_VAL_SPLIT,
+    early_stopping: bool = True,
+    patience: int = DEFAULT_PATIENCE,
+    n_iter_min: int = DEFAULT_N_ITER_MIN,
+    n_iter_print: int = DEFAULT_N_ITER_PRINT,
+    seed: int = DEFAULT_SEED,
+    return_val_loss: bool = False,
+    nonlin: str = DEFAULT_NONLIN,
+    avg_objective: bool = False,
+) -> Any:
     # function to train a single output head
     # input check
     y = check_shape_1d_data(y)
     d = X.shape[1]
     input_shape = (-1, d)
     rng_key = random.PRNGKey(seed)
-    onp.random.seed(seed) # set seed for data generation via numpy as well
+    onp.random.seed(seed)  # set seed for data generation via numpy as well
 
     # get validation split (can be none)
-    X, y, X_val, y_val, val_string = make_val_split(X, y, val_split_prop=val_split_prop, seed=seed)
-    n = X.shape[0] # could be different from before due to split
+    X, y, X_val, y_val, val_string = make_val_split(
+        X, y, val_split_prop=val_split_prop, seed=seed
+    )
+    n = X.shape[0]  # could be different from before due to split
 
     # get output head
-    init_fun, predict_fun = OutputHead(n_layers_out=n_layers_out, n_units_out=n_units_out,
-                                       binary_y=binary_y, n_layers_r=n_layers_r,
-                                       n_units_r=n_units_r, nonlin=nonlin)
+    init_fun, predict_fun = OutputHead(
+        n_layers_out=n_layers_out,
+        n_units_out=n_units_out,
+        binary_y=binary_y,
+        n_layers_r=n_layers_r,
+        n_units_r=n_units_r,
+        nonlin=nonlin,
+    )
 
     # get functions
     if not binary_y:
         # define loss and grad
         @jit
-        def loss(params, batch, penalty):
+        def loss(params: dict, batch: jnp.ndarray, penalty: float) -> jnp.ndarray:

```

Comment:
As far as I know, params is always a list, not a dict (here and elsewhere in the code). Further, batch is technically a tuple of jnp arrays, not an array.